### PR TITLE
Override Oracle JDBC native-image metadata only when native image is used

### LINKED ISSUE #35402: Override Oracle JDBC native-image metadata only when native image is used
## Description
The `OracleMetadataOverrides` class contains build steps, which manage the `native-image` properties of the Oracle JDBC driver and particular resources. However, these build steps are also executed when the native image is not used. Is it intentional? 

## Context
In Keycloak, we need to exclude the Oracle JDBC driver artifacts(`com.oracle.database.jdbc:ojdbc11`,`com.oracle.database.nls:orai18n`) from distribution due to particular productization issues. It means customers need to manually put the mentioned JARs into the distribution in order to run an instance with the Oracle database. The `quarkus-jdbc-oracle` extension is still present.

I downloaded the latest Oracle JDBC Driver (23.2.0.0) JAR and included it in the distribution. However, during augmentation, this warning is shown:
```
WARN [io.quarkus.deployment.steps.ClassTransformingBuildStep] (build-19) Could not remove configured resources from the following artifacts as they were not found in the model: {com.oracle.database.jdbc:ojdbc11::jar=[oracle/nativeimage/CharacterSetFeature.class, oracle/nativeimage/Target_java_io_ObjectStreamClass.class]}
```

Regarding the code, it's clear what's happening with the relationship of the custom JDBC driver. I understand the approach with the custom JDBC driver, and the Oracle JDBC extension is not supported by Quarkus, but we do not have any other option :)

https://github.com/quarkusio/quarkus/blob/8d60118b8572d5861ee1e6091584ef91f9ecb748/extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java#L150-L160

Btw. the `oracle/nativeimage/Target_java_io_ObjectStreamClass.class` is not even part of the newest provided Oracle JDBC Driver 23.2.0.0.

## Summary

Is it possible to make particular build steps in the `OracleMetadataOverrides` class conditionally enabled only for the native images? Or is there any deeper logic/idea behind it?

 

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by mabartos**
Closes #35402

### CODE REVIEW

File: extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java [hunk 1]
```diff
@@ -142,24 +143,24 @@ void excludeOracleDirectives(BuildProducer<ExcludeConfigBuildItem> nativeImageEx
...
+    @BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
     NativeImageAllowIncompleteClasspathBuildItem naughtyDriver() {
         return new NativeImageAllowIncompleteClasspathBuildItem("quarkus-jdbc-oracle");
     }
 
-    @BuildStep
+    @BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
     RemovedResourceBuildItem overrideSubstitutions() {
         return new RemovedResourceBuildItem(ArtifactKey.fromString("com.oracle.database.jdbc:ojdbc11"),
                 Collections.singleton("oracle/nativeimage/Target_java_io_ObjectStreamClass.class"));
```
Comment (mabartos) [inline extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java:154]: The `oracle/nativeimage/Target_java_io_ObjectStreamClass.class` is not part of the most recent Oracle JDBC Driver 23.2.0.0. Should it be removed, or is there any support for backward-compatible JDBC drivers?

File: extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java [hunk 2]
```diff
@@ -45,7 +46,7 @@ public final class OracleMetadataOverrides {
      *
      * @param reflectiveClass buildItem producer
      */
-    @BuildStep
+    @BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
     void build(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
```
Comment (gsmet) [inline extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java:50]: I would expect these to be executed only when in native as `ReflectiveClassBuildItem` is only consumed in native.

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks. I added a few comments.

I don't think we want to make all of them specifically marked for native only as most of them will automatically be executed only for native. But I agree it would be better to add it for the methods producing `RemovedResourceBuildItem` which are executed always.

File: extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java [hunk 3]
```diff
@@ -142,24 +143,18 @@ void excludeOracleDirectives(BuildProducer<ExcludeConfigBuildItem> nativeImageEx
...
+    @BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
     NativeImageAllowIncompleteClasspathBuildItem naughtyDriver() {
         return new NativeImageAllowIncompleteClasspathBuildItem("quarkus-jdbc-oracle");
     }
 
-    @BuildStep
-    RemovedResourceBuildItem overrideSubstitutions() {
-        return new RemovedResourceBuildItem(ArtifactKey.fromString("com.oracle.database.jdbc:ojdbc11"),
-                Collections.singleton("oracle/nativeimage/Target_java_io_ObjectStreamClass.class"));
-    }
```
Comment (mabartos) [inline extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java:154]: The `oracle/nativeimage/Target_java_io_ObjectStreamClass.class` is not part of the most recent Oracle JDBC Driver 23.2.0.0. Should it be removed, or is there any support for backward-compatible JDBC drivers?

@gsmet @Sanne WDYT?

**Review by mabartos — COMMENTED** [review]
@gsmet @Sanne The PR is ready to review. Just need to clarify one thing below.

Is there any chance to backport these changes to Quarkus 3.2.x? 

Thanks

**Comment by yrodiere** [comment]
> I don't think we want to make all of them specifically marked for native only as most of them will automatically be executed only for native. But I agree it would be better to add it for the methods producing `RemovedResourceBuildItem` which are executed always.

Well considering the whole class is about native image, we could just slap a `@BuildSteps(onlyIf = NativeOrNativeSourcesBuild.class)` at the class level: that would work and we'd never have to worry about it anymore.

But I suppose targeting just the few methods that matter like @gsmet requested works too :shrug: 

(No no I'm not trying to sell [my](https://github.com/quarkusio/quarkus/commit/c2eecd749d12f9bdc01209e210a37672b47ed3e0) lovely `@BuildSteps` annotation :D)

**Comment by mabartos** [comment]
@yrodiere Yes, also a possible solution.

@gsmet @Sanne Any preference?

**Comment by gsmet** [comment]
@mabartos sorry for the delay. I think @yrodiere 's suggestion of using `@BuildSteps` at the top of the class and drop the method annotations makes the most sense.

**Comment by mabartos** [comment]
@gsmet Done

**Review by yrodiere — APPROVED** [review]
LGTM.

I also checked and [this change](https://github.com/quarkusio/quarkus/pull/35403/files#r1299933629) should work on Quarkus 3.2/3.3 as well, since those also use oracle-jdbc 23.2.0.0.

File: extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java (hunk 3 above)
Comment (mabartos) [reply extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java:154]: @yrodiere ok

If possible, it'd be great to backport it to Quarkus 3.2.

**Comment by gsmet** [comment]
@yrodiere so we need to backport it, right?

**Comment by yrodiere** [comment]
> @yrodiere so we need to backport it, right?

#35402 seems to affect 3.2/3.3 as well, so I'd say we do, yes.

File: extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java [hunk 4]
```diff
@@ -34,6 +35,7 @@
  * require it, so this would facilitate the option to revert to the older version in
  * case of problems.
  */
+@BuildStep(onlyIf = NativeOrNativeSourcesBuild.class)
```
Comment (gsmet) [inline extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java:38]:
```suggestion
@BuildSteps(onlyIf = NativeOrNativeSourcesBuild.class)
```

**Review by gsmet — DISMISSED** [review]

File: extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java (hunk 4 above)
Comment (mabartos) [reply extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleMetadataOverrides.java:38]: @gsmet Oops, you're right

**Pull request merged**